src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/DirectoryBasedStoreProvider.cs

```diff
@@ -0,0 +1,400 @@
+                                // adding a public+private pair, continue with the add.
+                                //
+                                // So, therefore, if we aren't adding a private key, or already have one,
+                                // we don't need to do anything.
+                                return;
+                            }
+                        }
+                    }
+                }
+
+                // This may well be the first time that we've added something to this store.
+                Directory.CreateDirectory(_storePath);
+
+                string thumbprint = copy.Thumbprint;
+                bool findOpenSlot;
+
+                // The odds are low that we'd have a thumbprint colission, but check anyways.
+                string existingFilename = FindExistingFilename(copy, _storePath, out findOpenSlot);
+
+                if (existingFilename != null)
+                {
+                    bool dataExistsAlready = false;
+
+                    // If the file on disk is just a public key, but we're trying to add a private key,
+                    // we'll want to overwrite it.
+                    if (copy.HasPrivateKey)
+                    {
+                        try
+                        {
+                            using (X509Certificate2 fromFile = new X509Certificate2(existingFilename))
+                            {
+                                if (fromFile.HasPrivateKey)
+                                {
+                                    // We have a private key, the file has a private key, we're done here.
+                                    dataExistsAlready = true;
+                                }
+                            }
+                        }
+                        catch (CryptographicException)
+                        {
+                            // We can't read this file anymore, but a moment ago it was this certificate,
+                            // so go ahead and overwrite it.
+                        }
+                    }
+                    else
+                    {
+                        // If we're just a public key then the file has at least as much data as we do.
+                        dataExistsAlready = true;
+                    }
+
+                    if (dataExistsAlready)
+                    {
+                        // The file was added but our collection hasn't resynced yet.
+                        // Set _certificates to null to force a resync.
+                        _certificates = null;
+                        return;
+                    }
+                }
+
+                string destinationFilename;
+
+                if (existingFilename != null)
+                {
+                    destinationFilename = existingFilename;
+                }
+                else if (findOpenSlot)
+                {
+                    destinationFilename = FindOpenSlot(thumbprint);
```

Comment:
Do we need to be concerned about another process racing to find a slot and ending up with the same one?